src/dawn/engine_dawn.cc

```diff
@@ -249,15 +457,33 @@ Result EngineDawn::Initialize(EngineConfig* config,
 }
 
 Result EngineDawn::CreatePipeline(::amber::Pipeline* pipeline) {
+  if (!device_) {
+    return Result("Dawn::SetShader: device is not created");
+  }
+  std::unordered_map<ShaderType, ::dawn::ShaderModule, CastHash<ShaderType>>
+      module_for_type;
+  ::dawn::ShaderModuleDescriptor descriptor;
+  descriptor.nextInChain = nullptr;
+
   for (const auto& shader_info : pipeline->GetShaders()) {
-    Result r = SetShader(shader_info.GetShaderType(), shader_info.GetData());
-    if (!r.IsSuccess())
-      return r;
+    ShaderType type = shader_info.GetShaderType();
+    const std::vector<uint32_t>& code = shader_info.GetData();
+    descriptor.code = code.data();
+    descriptor.codeSize = uint32_t(code.size());
+
+    auto shader = device_->CreateShaderModule(&descriptor);
+    if (!shader) {
+      return Result("Dawn::SetShader: failed to create shader");
+    }
+    if (module_for_type.count(type)) {
+      return Result("Dawn::SetShader: module for type already exists");
+    }
+    module_for_type[type] = shader;
   }
 
   switch (pipeline->GetType()) {
     case PipelineType::kCompute: {
-      auto module = module_for_type_[kShaderTypeCompute];
+      auto module = module_for_type[kShaderTypeCompute];
```

Comment:
Hm. Now that I read this again I'd recommend making this "auto&"  so we avoid unnecessary copies.